### PR TITLE
Feature/tox

### PR DESCRIPTION
**Pull request by UltimateLobster**

### CODE REVIEW

File: .github/workflows/update-readme.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      - name: Update Readme with Coverage Html
+        run: |
+          sed -i '/<!-- Pytest Coverage Comment:Begin -->/,/<!-- Pytest Coverage Comment:End -->/c\<!-- Pytest Coverage Comment:Begin -->\n\${{ steps.coverageComment.outputs.coverageHtml }}\n${{ steps.coverageComment.outputs.summaryReport }}\n<!-- Pytest Coverage Comment:End -->' ./README.md
+
+      - name: Commit & Push changes to Readme
+        uses: actions-js/push@master
+        with:
+          message: Automatic update coverage on Readme [skip actions]
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          branch: ${{ github.ref }}
```
Comment (Avivsalem) [inline .github/workflows/update-readme.yml:39]: newline at the end of file

Comment (Avivsalem) [inline .github/workflows/update-readme.yml:39]: newline at the EOF is missing

File: baseservice/device_reader_service.py [hunk 2]
```diff
@@ -20,7 +20,7 @@ def __init__(self, *,
                  read_timeout: float = 5,
                  max_batch_read_count: int = 1,
                  wait_for_batch_count: bool = False,
-                 **kwargs):
+                 **kwargs: Any):
```
Comment (Avivsalem) [inline baseservice/device_reader_service.py:23]: Any? not Dict?

File: baseservice/device_reader_service.py [hunk 3]
```diff
@@ -42,19 +42,26 @@ def __init__(self, *,
...
-            batch = []
+            batch: List[Tuple[InputDevice, Message, Dict[str, Any]]] = []
 
             # read first message with _read_timeout anyway
-            message, device_headers = transaction_scope.read_message(timeout=self._read_timeout)
-            if message is not None:
-                batch.append((self._aggregate_input_device.last_read_device, message, device_headers))
+            message_result = transaction_scope.read_message(timeout=self._read_timeout)
+            if InputTransactionScope.is_non_empty_message(message_result):
+                message, device_headers = message_result
```
Comment (Avivsalem) [inline baseservice/device_reader_service.py:61]: wtf is this?

File: baseservice/device_reader_service.py [hunk 4]
```diff
@@ -68,20 +75,24 @@ def _server_loop(self, cancellation_token: threading.Event):
                 else:
                     timeout = 0  # if not wait_for_batch, try to read another message without waiting at all
 
-                message, device_headers = transaction_scope.read_message(timeout=timeout)
-                if message is None:
-                    break  # no more messages to read
+                result = transaction_scope.read_message(timeout=timeout)
+                if InputTransactionScope.is_non_empty_message(result):
+                    message, device_headers = result
```
Comment (Avivsalem) [inline baseservice/device_reader_service.py:80]: i don't like this (i know it's a duplicate comment on the review. but i don't like this twice...)

File: baseservice/iodevices/base/input_devices.py [hunk 5]
```diff
@@ -1,14 +1,19 @@
...
 
 from baseservice.iodevices.base.common import Message, DeviceHeaders
 from baseservice.utils import KwargsException, StatefulListIterator
 
-from baseservice.iodevices.base.input_transaction import InputTransaction, NULL_TRANSACTION
+from baseservice.iodevices.base.input_transaction import InputTransaction
+
+EmptyReadMessageResult: TypeAlias = Tuple[None, None, None]
+NonEmptyMessageResult: TypeAlias = Tuple[Message, DeviceHeaders, InputTransaction]
+ReadMessageResult = Union[NonEmptyMessageResult, EmptyReadMessageResult]
```
Comment (Avivsalem) [inline baseservice/iodevices/base/input_devices.py:13]: i don't mind... but why did we need this?

File: baseservice/iodevices/base/input_devices.py [hunk 6]
```diff
@@ -60,20 +65,31 @@ def read_message(self,
         :return: a tuple of (Message, DeviceHeaders, Transaction) or EMPTY_RESULT if no message was available.
         the device headers, can contain extra information about the device that returned the message
         """
-        message, device_headers, transaction = self._read_message(timeout=timeout,
-                                                                  with_transaction=with_transaction)
-        if message is not None:
+        result = self._read_message(timeout=timeout, with_transaction=with_transaction)
+        if self.is_non_empty_message_result(result):
```
Comment (Avivsalem) [inline baseservice/iodevices/base/input_devices.py:69]: o.k. we really need to find an alternative for this...
you must explain why is this like that.

File: baseservice/iodevices/base/input_devices.py [hunk 7]
```diff
@@ -60,20 +65,31 @@ def read_message(self,
...
+            message, device_headers, transaction = result
             device_headers = device_headers or {}
             device_headers.setdefault(self.INPUT_DEVICE_NAME_HEADER, self.name)
 
-        if not with_transaction:
-            transaction = NULL_TRANSACTION
+            if not with_transaction:
+                from baseservice.iodevices.base.null_transaction import NULL_TRANSACTION
```
Comment (Avivsalem) [inline baseservice/iodevices/base/input_devices.py:75]: why the import here?

File: baseservice/iodevices/base/input_output_devices.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
+from abc import ABCMeta
```
Comment (Avivsalem) [inline baseservice/iodevices/base/input_output_devices.py:1]: nononono... why do we need this?! I really wanted to separate input and output devices on this project

File: baseservice/iodevices/base/input_transaction.py [hunk 9]
```diff
@@ -1,30 +1,37 @@
...
 from threading import Event
-from typing import Optional, Tuple, List, Union, TYPE_CHECKING
+from types import TracebackType
+from typing import Optional, Tuple, List, Union, TypeVar, Type, overload, Generic, TYPE_CHECKING
+
+from typing_extensions import TypeAlias, TypeGuard
 
-from baseservice.iodevices.base.common import Message, DeviceHeaders
 if TYPE_CHECKING:
+    from baseservice.iodevices.base.common import Message, DeviceHeaders
```
Comment (Avivsalem) [inline baseservice/iodevices/base/input_transaction.py:9]: why only on type checking?

File: baseservice/iodevices/base/input_transaction.py [hunk 10]
```diff
@@ -1,30 +1,37 @@
...
+class InputTransaction(Generic[TInputDevice], metaclass=ABCMeta):
     """
     this is the base class for input transaction objects.
 
     after reading a message from the device (using 'with_transaction=True') the reader should commit/rollback
     the transaction, to signal the device if the message is done processing or not
     """
 
-    def __init__(self, device: Optional['InputDevice'] = None):
+    Self = TypeVar("Self", bound='InputTransaction')
```
Comment (Avivsalem) [inline baseservice/iodevices/base/input_transaction.py:23]: why is that important? also. very bad name

File: baseservice/iodevices/base/input_transaction.py [hunk 11]
```diff
@@ -37,10 +44,22 @@ def finished(self) -> bool:
         """
         return self._finished.is_set()
 
-    def __enter__(self):
+    def __enter__(self: Self) -> Self:
         return self
 
-    def __exit__(self, exc_type, exc_val, exc_tb):
+    @overload
+    def __exit__(self, exc_type: None, exc_val: None, exc_tb: None) -> None:
```
Comment (Avivsalem) [inline baseservice/iodevices/base/input_transaction.py:51]: explain?

File: baseservice/iodevices/base/input_transaction.py [hunk 12]
```diff
@@ -76,26 +95,32 @@ def rollback(self):
...
     @abstractmethod
-    def _rollback(self):
+    def _rollback(self) -> None:
         """
         this method should be implemented by child classes to actually perform the rollback
         """
         pass
 
 
+EmptyInputTransactionScopeMessage: TypeAlias = Tuple[None, None]
```
Comment (Avivsalem) [inline baseservice/iodevices/base/input_transaction.py:112]: AHHHHHHHHHHHHHHHHH!!!!

File: baseservice/iodevices/base/input_transaction.py [hunk 13]
```diff
@@ -105,7 +130,7 @@ def __init__(self, device: Optional['InputDevice'] = None, with_transaction: boo
         self._with_transaction = with_transaction
         self._transactions: List[InputTransaction] = []
 
-    def read_message(self, timeout: Optional[float] = 0) -> Union[Tuple[Message, DeviceHeaders], Tuple[None, None]]:
+    def read_message(self, timeout: float = 0) -> InputTransactionScoptMessage:
```
Comment (Avivsalem) [inline baseservice/iodevices/base/input_transaction.py:133]: why did you remove the Optional? timeout can be None (for block forever)

File: baseservice/iodevices/in_memory_device/in_memory_device_manager.py [hunk 14]
```diff
@@ -2,45 +2,46 @@
...
                                         NULL_TRANSACTION,
                                         ReadMessageResult,
                                         EMPTY_RESULT)
+from baseservice.iodevices.base.input_output_devices import InputOutputDevice, InputOutputDeviceManager
 
 MESSAGE_TIMESTAMP_HEADER = 'message_timestamp'
 
 
-class InMemoryDevice(InputDevice, OutputDevice):
+class InMemoryDevice(InputOutputDevice):
```
Comment (Avivsalem) [inline baseservice/iodevices/in_memory_device/in_memory_device_manager.py:20]: don't like this. why is that necessary?

File: baseservice/iodevices/short_circuit_device_wrapper/short_circuit_input_device.py [hunk 15]
```diff
@@ -15,7 +13,7 @@ def __init__(self,
         ShortCircuitDeviceBase.__init__(self, short_circuit_fail_count, short_circuit_time)
         self._inner_device = inner_device
 
-    def _read_message(self, timeout: Optional[float] = 0, with_transaction: bool = True) -> ReadMessageResult:
+    def _read_message(self, timeout: float = 0, with_transaction: bool = True) -> ReadMessageResult:
```
Comment (Avivsalem) [inline baseservice/iodevices/short_circuit_device_wrapper/short_circuit_input_device.py:16]: again. should be Optional

File: baseservice/utils.py [hunk 16]
```diff
@@ -92,38 +93,49 @@ def __init__():
     a.z # raises AttributeError (z was not set on class). once it's set, then value is the init value for each thread
     a.w # 2 for this thread, 1 for any other thread (because of the default init value)
     """
+    __NO_INIT_VALUE = None
 
-    def __init__(self, init_value: Optional[TValueType] = ...):
+    def __init__(self, init_value: Optional[TValueType] = __NO_INIT_VALUE):
```
Comment (Avivsalem) [inline baseservice/utils.py:98]: what's wrong with the ellipsis?

File: tests/__init__.py [hunk 17]
```diff
@@ -1 +0,0 @@
-
```
Comment (Avivsalem) [inline tests/__init__.py:1]: what do we do when we want to load extra requirements from files? don't we have to have setup.py?

**Review by Avivsalem — CHANGES_REQUESTED** [review]
All the mess with the None read message type, makes me wonder if it's not better to return dataclass instead of tuple from read_message. that way it can be optional.

what do you think?

also, you need to explain some things to me (like, what's "overload" and what is it good for. or TypeGuard...)

File: baseservice/utils.py (hunk 16 above)
Comment (UltimateLobster) [reply baseservice/utils.py:98]: Not that it's important anymore but usually sentinels are represented with object() instead of elipssis (which i think only pydantic does)
